### PR TITLE
pkg/loadbalancer: Optimize L3n4Addr.Hash for performance

### PR DESCRIPTION
**Pull request by gandro**
This commit optimizes the L3n4Addr.Hash function. It is used as a key in
Golang maps and therefore needs to be unique, but it does not need to actually
be a secure hash.

The previous implementation was slow issues because it relied on defer
statements, reflection (due to the use of `%+v`), and SHA256. This
function is used during service lookup in Hubble's hot path, therefore
this optimization should reduce the overhead of Hubble:

image

Before this commit:

    BenchmarkL3n4Addr_Hash_IPv4-8         	  608378	      1856 ns/op
    BenchmarkL3n4Addr_Hash_IPv6_Short-8   	  620024	      2080 ns/op
    BenchmarkL3n4Addr_Hash_IPv6_Long-8    	  535290	      2168 ns/op

After this commit:

    BenchmarkL3n4Addr_Hash_IPv4-8         	15079537	        81.1 ns/op
    BenchmarkL3n4Addr_Hash_IPv6_Short-8   	 8348496	       149 ns/op
    BenchmarkL3n4Addr_Hash_IPv6_Long-8    	 4463416	       259 ns/op

### CODE REVIEW

**Review by brb — APPROVED** [review]
Nice! Do you have a flamegraph for "after"?

**Comment by gandro** [comment]
> Nice! Do you have a flamegraph for "after"?

This is the flamegraph for just the service lookup afterwards. Most time is still spent in Hash, but now it's mostly memory allocation.

image

**Review by borkmann — APPROVED** [review]

File: pkg/loadbalancer/loadbalancer.go [hunk 1]
```diff
@@ -475,17 +475,21 @@ func (a *L3n4Addr) DeepCopy() *L3n4Addr {
...
-	}()
-
-	str := []byte(fmt.Sprintf("%+v", a))
-	return fmt.Sprintf("%x", sha512.Sum512_256(str))
+	b = append(b, '|')
+	b = strconv.AppendUint(b, uint64(a.Port), 10)
+	b = append(b, '|')
+	b = strconv.AppendUint(b, uint64(a.Scope), 10)
+
+	return string(b)
```
Comment (borkmann) [inline pkg/loadbalancer/loadbalancer.go:497]: One more thing, how does this relate to L3n4AddrID? E.g. https://github.com/cilium/cilium/blob/2b9485df05ec24fe051684ce992a82dabd0b9e56/pkg/maps/lbmap/lbmap.go#L563 is calling the Hash() from an L3n4AddrID type where the Sprintf() was on %+v.

Comment (gandro) [reply pkg/loadbalancer/loadbalancer.go:497]: Calling `fe.Hash()` (where `fe` is of type `L3n4AddrID`) is equivalent to calling `fe.L3n4Addr.Hash()`. In other words, there is no inheritance. The `%+v` part was still only printing the contents of `L3n4Addr`, the `ID` is and was omitted. 

Not sure that is intended, but this commit does not change that.

**Review by aditighag — APPROVED** [review]
🚀

File: pkg/loadbalancer/loadbalancer.go [hunk 2]
```diff
@@ -475,17 +475,21 @@ func (a *L3n4Addr) DeepCopy() *L3n4Addr {
...
 }
 
-// Hash calculates L3n4Addr's internal SHA256Sum.
+// Hash calculates a unique string of the L3n4Addr e.g for use as a key in maps
 func (a L3n4Addr) Hash() string {
+	// Note: 32 bytes is not enough for long IPv6 addresses, but it is cheaper
+	// to reallocate on overflow than to check if a.IP is IPv6
+	b := make([]byte, 0, 32)
+
+	b = append(b, a.IP.String()...)
```
Comment (aanm) [inline pkg/loadbalancer/loadbalancer.go]: why don't we do something like:
```
ip := a.IP.String()
if bytes.IndexRune(ip, ':') < 0 {
    b = make([]byte, 0, maxV4 /*(=32)*/)
} else {
    b = make([]byte, 0, maxV6 /*(=39? + 5 + sizeOfScope)*/)
}

File: pkg/loadbalancer/loadbalancer.go [hunk 3]
```diff
@@ -475,17 +475,21 @@ func (a *L3n4Addr) DeepCopy() *L3n4Addr {
 	}
 }
 
-// Hash calculates L3n4Addr's internal SHA256Sum.
+// Hash calculates a unique string of the L3n4Addr e.g for use as a key in maps
```
Comment (kaworu) [inline pkg/loadbalancer/loadbalancer.go:478]: At that point would it make sense to rename the function from `Hash` to something like `Key`?

File: pkg/loadbalancer/loadbalancer.go (hunk 1 above)
Comment (kaworu) [reply pkg/loadbalancer/loadbalancer.go:497]: `L3n4AddrID.Hash` is simply `L3n4Addr.Hash` and thus the `ID` field will not be used. If that is an issue, we probably want to define `L3n4AddrID.Hash` as `L3n4Addr.Hash()|ID` or something along those lines.

File: pkg/loadbalancer/loadbalancer.go [hunk 4]
```diff
@@ -475,17 +475,21 @@ func (a *L3n4Addr) DeepCopy() *L3n4Addr {
...
 func (a L3n4Addr) Hash() string {
+	// Note: 32 bytes is not enough for long IPv6 addresses, but it is cheaper
+	// to reallocate on overflow than to check if a.IP is IPv6
+	b := make([]byte, 0, 32)
```
Comment (kaworu) [inline pkg/loadbalancer/loadbalancer.go]: FWIW Looks like we only need to raise to `36` bytes to spare an extra allocation in `BenchmarkL3n4Addr_Hash_IPv6_Long`.

**EDIT**: my bad, the IPv6 in `BenchmarkL3n4Addr_Hash_IPv6_Long` is not the longest. An IPv6 that cannot be shortened require `46` bytes for `b`.

File: pkg/loadbalancer/loadbalancer.go (hunk 2 above)
Comment (gandro) [reply pkg/loadbalancer/loadbalancer.go]: ~This seems to do the trick without a measurable overhead. Will change.~

File: pkg/loadbalancer/loadbalancer.go (hunk 4 above)
Comment (gandro) [reply pkg/loadbalancer/loadbalancer.go]: Yes. I'll a benchmark for the worst-case as well.

File: pkg/loadbalancer/loadbalancer.go (hunk 2 above)
Comment (gandro) [reply pkg/loadbalancer/loadbalancer.go]: Actually, after some more benchmarking, simply relying on reallocation is still faster in most cases than trying to be smart about it:

---

Fastest / Most imprecise:

``` go
const lenIPv4 = 15
const lenProto = 1
const lenPort = 6
const lenScope = 2

// Note: The capacity might not be enough for long IPv6 addresses, but it is cheaper
// to reallocate on overflow than to check the length of a.IP
b := make([]byte, 0, lenIPv4+lenProto+lenPort+lenScope)
```

```
BenchmarkL3n4Addr_Hash_IPv4-8         	16097442	        70.7 ns/op
BenchmarkL3n4Addr_Hash_IPv6_Short-8   	 9431970	       125 ns/op
BenchmarkL3n4Addr_Hash_IPv6_Long-8    	 5561029	       215 ns/op
BenchmarkL3n4Addr_Hash_IPv6_Max-8     	 3719448	       314 ns/op
```

---

Middle ground:

``` go
const lenIPv4 = 15
const lenIPv6 = 39
const lenProto = 1
const lenPort = 6
const lenScope = 2

var b []byte
ip := a.IP.String()
if strings.IndexRune(ip, ':') < 0 {
    b = make([]byte, 0, lenIPv4+lenProto+lenPort+lenScope)
} else {
    b = make([]byte, 0, lenIPv6+lenProto+lenPort+lenScope)
}
```

```
BenchmarkL3n4Addr_Hash_IPv4-8         	14471590	        80.9 ns/op
BenchmarkL3n4Addr_Hash_IPv6_Short-8   	 9139532	       132 ns/op
BenchmarkL3n4Addr_Hash_IPv6_Long-8    	 6556764	       204 ns/op
BenchmarkL3n4Addr_Hash_IPv6_Max-8     	 3862281	       313 ns/op
```

---

Slowest / Most-precise:

```go
const lenProto = 1
const lenPort = 6
const lenScope = 2

ip := a.IP.String()
b := make([]byte, 0, len(ip)+lenProto+lenPort+lenScope)
```

```
BenchmarkL3n4Addr_Hash_IPv4-8         	12182581	        97.3 ns/op
BenchmarkL3n4Addr_Hash_IPv6_Short-8   	 7979997	       147 ns/op
BenchmarkL3n4Addr_Hash_IPv6_Long-8    	 5991332	       215 ns/op
BenchmarkL3n4Addr_Hash_IPv6_Max-8     	 3743278	       317 ns/op
```

File: pkg/loadbalancer/loadbalancer.go (hunk 3 above)
Comment (gandro) [reply pkg/loadbalancer/loadbalancer.go:478]: For ease of potential back-porting, I'll leave this as is for now.

**Comment by gandro** [comment]
Pushed an additional benchmark and made the length calculation a bit more tight (24 instead of 32 bytes by default). See also https://github.com/cilium/cilium/pull/14617#discussion_r557568344

**Comment by gandro** [comment]
test-me-please

**Review by christarazi — APPROVED** [review]
:rocket:

**Comment by gandro** [comment]
Net-Next hit #14598
https://jenkins.cilium.io/job/Cilium-PR-K8s-1.13-net-next/460/
Two errors in the image build that look spurious and unrelated:
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Kernel/4448/
https://jenkins.cilium.io/job/Cilium-PR-K8s-1.20-kernel-4.9/408/

**Comment by gandro** [comment]
retest-net-next

**Comment by gandro** [comment]
retest-kernel-4.9

**Comment by gandro** [comment]
retest-4.9

**Comment by gandro** [comment]
retest-4.19

File: pkg/loadbalancer/loadbalancer.go [hunk 5]
```diff
@@ -475,17 +475,26 @@ func (a *L3n4Addr) DeepCopy() *L3n4Addr {
 	}
 }
 
-// Hash calculates L3n4Addr's internal SHA256Sum.
+// Hash calculates a unique string of the L3n4Addr e.g for use as a key in maps
 func (a L3n4Addr) Hash() string {
+	const lenIPv4 = 15
```
Comment (aanm) [inline pkg/loadbalancer/loadbalancer.go:480]: `const lenIPv4 = 16` // len("255.255.255.255|")

File: pkg/loadbalancer/loadbalancer.go [hunk 6]
```diff
@@ -475,17 +475,26 @@ func (a *L3n4Addr) DeepCopy() *L3n4Addr {
...
+	const lenProto = 1
+	const lenPort = 6
+	const lenScope = 2
```
Comment (aanm) [inline pkg/loadbalancer/loadbalancer.go:483]: `const lenScope = 1` // len("0") or len("1"), there isn't a "|" as part of the scope

**Review by kaworu — APPROVED** [review]
Thanks LGTM!

File: pkg/loadbalancer/loadbalancer.go (hunk 6 above)
Comment (gandro) [reply pkg/loadbalancer/loadbalancer.go:483]: The sum should be correct. I did not add the `|` to the IP address, and added it to the scope instead

**Comment by gandro** [comment]
net-next hit #14598 again, as well as an error similar to #12690 - but in a different context: https://jenkins.cilium.io/job/Cilium-PR-K8s-1.13-net-next/468/testReport/junit/Suite-k8s-1/13/K8sServicesTest_Checks_service_across_nodes_Tests_NodePort_BPF_Tests_with_direct_routing_With_host_policy_Tests_NodePort/

**Comment by gandro** [comment]
retest-net-next

**Pull request merged**